### ISSUE TITLE
Spurious "unreachable code" warning with confused prefixes

### ISSUE THREAD

**Issue by scabug** [scala/bug#9220]
The following,

```scala
trait Command
object Command {
  sealed trait Execution extends Command
}

case class Buzz() extends Command.Execution
case class Door() extends Command.Execution

def foo(cmd: Command.Execution) = cmd match {
  case x @ (_: Buzz) => ???
  case x @ (_: Door) => ???
}
```

results in the following spurious unreachable code warning,

```scala
[error]  bug.scala:11: unreachable code
[error]       case x @ (_: Door) => ???
[error]                             ^
```

Unnesting the `Execution` trait from the `Command` companion object results in compilation without the complaint.

**Comment by OlivierBlanvillain** [comment]
This issue seams to be fixed in 2.12.

**Comment by SethTisue** [comment]
I'm unable to reproduce it in 2.10 or 2.11 either, though. That has me wondering if some special compiler flag is required, or what. (I do have `-Xlint` enabled.)

**Comment by Jasper-M** [comment]
@SethTisue Strange, I can reproduce it without any special flags in every version I tried except for 2.12.x

**Comment by SethTisue** [comment]
oh, never mind, I can reproduce it with Jason's version of the code.

**Comment by SethTisue** [comment]
and yup, fixed in 2.12
